### PR TITLE
Add option to download input files using a local MinIO server

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context

Add functionality to download extract files from AWS S3 to a local MinIO server that acts as a "local S3 bucket" . This update improves the performance of the app by reducing the number of data pulls / requests sent to AWS S3 (potentially reducing costs down the line as the app gets more use) and avoiding repeated downloads of extract files.

Below are some key takeaways:

1. The **default** behavior of CLI command `run-diff` is to _not_ download extract files (i.e., Docker containers access input files from AWS S3 given the appropriate credentials as environment variables).
2. To download input files, [pass the flag `--download-files` when invoking the CLI command `run-diff`](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/49/files#diff-8f04cb2bf1d6814d94e2907a667241569a9004024a32a4af79c21fd098adbc1cR152-R171).
3. Instructions for using the workflow with MinIO are provided in [README | Running a Local MinIO Server](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-353-download-extract-files/README.md#running-a-local-minio-server).
4. This update introduces [five (5) new, optional environment variables](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/49/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R125-R129). ~Only one of these new environment variables must be explicitly set in the `.env` file (`MINIO_S3_LOCAL_STORAGE`); [default values are set for the four other environment variables in `abdiff.config`](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/49/files#diff-a0cfcca0f834ffcd25602389082872c5dd2786e3abd9aa3f8727780338561a34R33-R47).~
   **UPDATED**: [Three of these new environment variables must be explicitly set in the `.env` file](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/49#discussion_r1833038545).

### How can a reviewer manually see the effects of these changes?

No additional unit tests have been added as these changes are somewhat complex to model via unit tests. However, example commands (with screenshots of the output) should be sufficient for purposes of review. 

1. Follow the instructions in [README | Running a Local MinIO Server](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-353-download-extract-files/README.md#running-a-local-minio-server) to launch a local MinIO server with a Docker container. **Make sure to set Dev AWS credentials in your `.env` file.**

2. Run the following CLI commands in order:
   * Initiate a job.
      ```shell
      pipenv run abdiff --verbose init-job \
         -d output/jobs/test-minio \
        -m "test job with MinIO support" \
        -a 395e612 \
        -b cf1024c
      ```
   * Run a diff:
      ```shell
      pipenv run abdiff --verbose run-diff \
         -d output/jobs/test-minio \
         -i "s3://timdex-extract-dev-222053980223/libguides/libguides-2024-06-03-full-extracted-records-to-index.xml,s3://timdex-extract-dev-222053980223/dspace/dspace-2024-03-06-daily-extracted-records-to-index.xml" \
         -m "running diff on two input files with MinIO support" \
         --download-files
      ```
3. Review the **output** (redacted to focus on `download_input_files` and `run_ab_transforms`:
```shell
2024-11-07 12:02:17,617 INFO abdiff.cli.main() line 44: Logger 'root' configured with level=DEBUG
2024-11-07 12:02:17,617 INFO abdiff.cli.main() line 45: Running process
2024-11-07 12:02:17,618 INFO abdiff.core.init_run.init_run() line 24: Run directory created: output/jobs/test-minio/runs/2024-11-07_17-02-17
2024-11-07 12:02:17,747 INFO abdiff.extras.minio.download_input_files.download_input_files() line 49: Input file: 1 / 2: 's3://timdex-extract-dev-222053980223/libguides/libguides-2024-06-03-full-extracted-records-to-index.xml' available locally for transformation.
2024-11-07 12:02:17,751 INFO abdiff.extras.minio.download_input_files.download_input_files() line 49: Input file: 2 / 2: 's3://timdex-extract-dev-222053980223/dspace/dspace-2024-03-06-daily-extracted-records-to-index.xml' available locally for transformation.
2024-11-07 12:02:17,751 INFO abdiff.extras.minio.download_input_files.download_input_files() line 59: Available input files: 2, missing input files: 0.
```

4. Run the following command that includes an **invalid** input file.
   ```shell
   pipenv run abdiff --verbose run-diff \
      -d output/jobs/test-minio \
      -i "s3://timdex-extract-dev-222053980223/libguides/libguides-2024-06-03-full-extracted-records-to-index.xml,s3://timdex-extract-dev-222053980223/dspace/dspace-2024-03-06-daily-extracted-records-to-index.xml,s3://timdex-extract-dev-222053980223/libguides/libguides-2024-11-07-daily-extracted-records-to-index.xml" \
      -m "running diff on two input files with MinIO support" \
      --download-files
   ```

   This will result in the following logs and error:
```
2024-11-07 12:03:27,559 INFO abdiff.cli.main() line 44: Logger 'root' configured with level=DEBUG
2024-11-07 12:03:27,559 INFO abdiff.cli.main() line 45: Running process
2024-11-07 12:03:27,561 INFO abdiff.core.init_run.init_run() line 24: Run directory created: output/jobs/test-minio/runs/2024-11-07_17-03-27
2024-11-07 12:03:27,683 INFO abdiff.extras.minio.download_input_files.download_input_files() line 49: Input file: 1 / 3: 's3://timdex-extract-dev-222053980223/libguides/libguides-2024-06-03-full-extracted-records-to-index.xml' available locally for transformation.
2024-11-07 12:03:27,688 INFO abdiff.extras.minio.download_input_files.download_input_files() line 49: Input file: 2 / 3: 's3://timdex-extract-dev-222053980223/dspace/dspace-2024-03-06-daily-extracted-records-to-index.xml' available locally for transformation.
2024-11-07 12:03:28,816 INFO abdiff.extras.minio.download_input_files.download_input_files() line 55: Input file: 3 / 3: 's3://timdex-extract-dev-222053980223/libguides/libguides-2024-11-07-daily-extracted-records-to-index.xml' failed to download.
2024-11-07 12:03:28,816 INFO abdiff.extras.minio.download_input_files.download_input_files() line 59: Available input files: 2, missing input files: 1.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/Documents/repos/transmogrifier-ab-diff/abdiff/cli.py", line 175, in run_diff
    download_input_files(input_files_list)
  File "/Users/jcuerdo/Documents/repos/transmogrifier-ab-diff/abdiff/extras/minio/download_input_files.py", line 64, in download_input_files
    raise RuntimeError(  # noqa: TRY003
RuntimeError: 1 input file(s) failed to download.
``` 

### Includes new or updated dependencies?
YES

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/TIMX-353 

### Developer
- [X] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -54,3 +55,10 @@ black-apply: # Apply changes with 'black'
 
 ruff-apply: # Resolve 'fixable errors' with 'ruff'
 	pipenv run ruff check --fix .
+
+# Development commands
```
Comment (ghukill) [inline Makefile:59]: I'd propose updating to something like "Minio local S3 commands"

File: README.md [hunk 2]
```diff
@@ -15,6 +15,38 @@ Compare transformed TIMDEX records from two versions (A,B) of Transmogrifier.
 - To lint the repo: `make lint`
 - To run the app: `pipenv run abdiff --help`
 
+### Running a Local MinIO Server
+
+TIMDEX extract files from S3 (i.e., input files to use in transformations) can be downloaded to a local MinIO server hosted with a Docker container. [MinIO is an object storage solution that provides an Amazon Web Services S3-compatible API and supports all core S3 features](https://min.io/docs/minio/kubernetes/upstream/). Downloading extract files improves the runtime of a diff by reducing the number of requests sent to S3 and avoids repeated downloads of extract files.
```
Comment (ghukill) [inline README.md:20]: Can we add an additional sentence that utilizing local files also avoids AWS credentials timing out?

File: abdiff/cli.py [hunk 3]
```diff
@@ -148,7 +149,12 @@ def init_job(
     help="Message to describe Run.",
     default="Not provided.",
 )
-def run_diff(job_directory: str, input_files: str, message: str) -> None:
+@click.option(
+    "--download-files", is_flag=True, help="Pass to skip download of extract files"
```
Comment (ghukill) [inline abdiff/cli.py:153]: Does the help message need updating?  I would think it's something like,

> "Pass to download input files from AWS S3 to a local Minio S3 server for Transmogrifier to use."

File: abdiff/config.py [hunk 4]
```diff
@@ -25,6 +30,22 @@ def __getattr__(self, name: str) -> Any:  # noqa: ANN401
...
+        return self.MINIO_S3_CONTAINER_URL or "http://host.docker.internal:9000/"
+
+    @property
+    def minio_root_user(self) -> str:
+        return self.MINIO_ROOT_USER or "minioadmin"
+
+    @property
+    def minio_root_password(self) -> str:
+        return self.MINIO_ROOT_PASSWORD or "minioadmin"
+
```
Comment (ghukill) [inline abdiff/config.py:48]: +1 - liking the use of dynamic `Config` class properties

File: abdiff/config.py [hunk 5]
```diff
@@ -25,6 +30,22 @@ def __getattr__(self, name: str) -> Any:  # noqa: ANN401
         message = f"'{name}' not a valid configuration variable"
         raise AttributeError(message)
 
+    @property
+    def minio_s3_url(self) -> str:
+        return self.MINIO_S3_URL or "http://localhost:9000/"
+
+    @property
+    def minio_s3_container_url(self) -> str:
+        return self.MINIO_S3_CONTAINER_URL or "http://host.docker.internal:9000/"
```
Comment (ghukill) [inline abdiff/config.py:39]: Can these both have small notes explaining what they are?

Something like:
- `minio_s3_url`: "Host for ABDiff context (host machine) to connect to Minio."
- `minio_s3_container_url`: "Host for Transmogrifier Docker containers to connect to Minio."

File: abdiff/core/download_input_files.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+          file exists prior to any download.
+    """
+    s3_client = boto3.client(
+        "s3",
+        endpoint_url=CONFIG.minio_s3_url,
+        aws_access_key_id=CONFIG.minio_root_user,
+        aws_secret_access_key=CONFIG.minio_root_password,
+    )
+
+    for input_file in input_files:
```
Comment (ghukill) [inline abdiff/core/download_input_files.py:43]: Can we add some progress logging here, something like:

```python
for i, input_file in enumerate(input_files):
    ...
    ...
```

And then perhaps skip the other logging messages, and add one shared one at the end?  Something like:

```python
logger.info(f"Input file {i+1} / {len(input_files)}: '{input_file}' available locally for transformation.")
```

For very large runs, I think this consistent output would be helpful.  I suppose we lose the grain of skipped vs downloaded, but the ouput would be consistent.  Thoughts?

File: abdiff/core/download_input_files.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        logger.info(f"Downloading input file from {CONFIG.TIMDEX_BUCKET}: {input_file}")
+        copy_command = ["aws", "s3", "cp", input_file, "-"]
+        upload_command = [
+            "aws",
+            "s3",
+            "cp",
+            "--endpoint-url",
+            CONFIG.minio_s3_url,
+            "--profile",
+            "minio",
```
Comment (ghukill) [inline abdiff/core/download_input_files.py:57]: At first I bristled at having to create a "minio" AWS profile that this could use... but I've warmed up to it.  It certainly works well here, and furthermore, this profile then becomes something I could use outside of ABDiff explicitly if I wanted to manage my own minio data.  

I have a strong feeling that this local S3 / copy of TIMDEX S3, may have _large_ benefits as we get into more involved TIMDEX work.  I'm happy that we're getting some repetitions here utilizing it.  Perhaps we evolve the ergonomics as we go, but this feels great for now.

File: abdiff/core/download_input_files.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+from mypy_boto3_s3.client import S3Client
+
+from abdiff.config import Config
+
+logger = logging.getLogger(__name__)
+
+CONFIG = Config()
+
+
+def download_input_files(input_files: list[str]) -> None:
```
Comment (ghukill) [inline abdiff/core/download_input_files.py:15]: Commenting generally on this function.

It's possible / likely / happened, that I suggested to @jonavellecuerdo that we keep this simple for now and perform these downloads sequentially.  That said, observing that this would be ripe for threaded (concurrent) downloads.  It would speed this up _considerably_, particularly if we are downloading 1.6k+ files @ 50+ gb.

@ehanson8 , @jonavellecuerdo - would you prefer to create a ticket for add concurrency here?  or have that implemented now?

File: abdiff/core/download_input_files.py
Comment (ghukill) [inline abdiff/core/download_input_files.py:1]: On further reflection, I suppose I would like to see this `download_input_files.py` located in the `abdiff/extras/minio` folder.  It does not strike me as "core" functionality.

I think it's okay that we _reference_ a "non-core" function in the `run-diff` CLI command -- seen this in other applications, e.g. pandas or DuckDB -- but it might be nice to locate this code elsewhere.  

Thoughts @jonavellecuerdo?

File: abdiff/helpers/minio/docker-compose.yaml
Comment (ghukill) [inline abdiff/helpers/minio/docker-compose.yaml:1]: I hate to say it, but this Docker Compose file feels a little copy/pasta from some spike code and the Minio examples.  

Specifically, I don't think we need to `x-minio-common` and the `<<: *minio-common` later in the file.   And perhaps we don't need the `RELEASE.2024-10-29T16-01-48Z` tag, but could just use `latest`.

Could we just simplify to a more straight-forward compose file?

```
services:
  minio:
    image: quay.io/minio/minio:latest
    command: server --console-address ":9001" /mnt/data
    ports:
      - "9000:9000"  # API port
      - "9001:9001"  # Console port
    environment:
      MINIO_ROOT_USER: ${MINIO_ROOT_USER}
      MINIO_ROOT_PASSWORD: ${MINIO_ROOT_PASSWORD}
    healthcheck:
      test: [ "CMD", "mc", "ready", "local" ]
      interval: 5s
      timeout: 5s
      retries: 5
    volumes:
      - ${MINIO_S3_LOCAL_STORAGE}:/mnt/data
```

File: Makefile [hunk 9]
```diff
@@ -54,3 +55,10 @@ black-apply: # Apply changes with 'black'
...
+start-minio-server:
+	docker compose --env-file .env -f $(MINIO_COMPOSE_FILE) up -d
+
+stop-minio-server:
+	docker compose -f $(MINIO_COMPOSE_FILE) stop
```
Comment (ghukill) [inline Makefile:64]: I think perhaps this needs the `.env` file as well?

```
docker compose --env-file .env -f $(MINIO_COMPOSE_FILE) stop
```

**Review by ghukill — CHANGES_REQUESTED** [review]
Overall, I think this looks really good.  Nice work @jonavellecuerdo!  As we've discussed on huddles, I think this work is more in the "ergonomics" camp than raw functionality, which is often much harder in my opinion.

I was able to get it running successfully, performing a run via local files via my Minio instance.  Whoohoo!  

I have left a handful of comments, which tips into requesting some changes.  Not all are mandatory (requested), but a couple jump out:

- simplify Minio docker compose file
- move references to `abdiff/helpers` to `abdiff/extras`
- consider moving `download_input_files()` to `abdiff/extras/minio`

I also left a question about concurrent downloads, but truly happy to wait on that to keep this PR focused on functionality.

File: Makefile (hunk 1 above)
Comment (jonavellecuerdo) [reply Makefile:59]: Updated to: 

```text
####################################
# MinIO local S3 commands
####################################
```

File: README.md (hunk 2 above)
Comment (jonavellecuerdo) [reply README.md:20]: Let me know what you think of this revision:

> TIMDEX extract files from S3 (i.e., input files to use in transformations) can be downloaded to a local MinIO server hosted via a Docker container. [MinIO is an object storage solution that provides an Amazon Web Services S3-compatible API and supports all core S3 features](https://min.io/docs/minio/kubernetes/upstream/). The MinIO server acts as a "local S3 file system", allowing the app to access data on disk through an S3 interface. Since the MinIO server runs in a Docker container, it can be easily started when needed and stopped when not in use. Any data stored in the MinIO server will persist as long as the files exist in the directory specified for `MINIO_S3_LOCAL_STORAGE`.
>
> Downloading extract files improves the runtime of a diff by reducing the number of requests sent to S3 and avoids AWS credentials timing out. Once an extract file is stored in the local MinIO server, the app can access the data from MinIO for all future runs that include the extract file, avoiding repeated downloads of data used across multiple runs.

Comment (ghukill) [reply README.md:20]: Beautiful!  Thanks.

File: abdiff/cli.py (hunk 3 above)
Comment (jonavellecuerdo) [reply abdiff/cli.py:153]: Updated.

File: abdiff/core/download_input_files.py (hunk 6 above)
Comment (jonavellecuerdo) [reply abdiff/core/download_input_files.py:43]: Hmm, I agree that in the case of large runs, it would be hard to parse through the varying logs to identify files that were already in the S3 bucket (and downloads were skipped for). 🤔 I'm good with the change proposed above if @ehanson8 agrees with the change as well!

**Outside the scope of this PR:** It feels like a CLI command that informs / warns / reminds user to clean up their storage in the case of very large runs, but it seems like that CLI command would only know which directory to check via the `MINIO_S3_LOCAL_STORAGE` environment variable (i.e., if an initial diff with large number of input files was run where MinIO storage is set to Folder A and a second diff was planned where MinIO storage is set to Folder B, the CLI command wouldn't raise any warnings when checking Folder B). I might've brought this up in a call with @ghukill , but a safeguard of some sort to remind users of previously downloaded data feels like it'd be helpful. 🤔

File: README.md [hunk 10]
```diff
@@ -15,6 +15,38 @@ Compare transformed TIMDEX records from two versions (A,B) of Transmogrifier.
...
+   ```
+
+   The API is accessible at: http://127.0.0.1:9000.
+   The WebUI is accessible at: http://127.0.0.1:9001.
+
+4. On your browser, navigate to the WebUI and sign into the local MinIO server. Create a bucket in the local MinIO server named after the S3 bucket containing the TIMDEX extract files that will be used in the A/B Diff.
+
+5. Proceed with A/B Diff CLI commands as needed!
+
+Once a diff run is complete, you can stop the local MinIO server using the Makefile command: `make stop-minio-server`. If you're planning to run another diff using the same files -- good news! All you have to do is restart the local MinIO server. Your data will persist as long as the files exist in the directory you specified for `MINIO_S3_LOCAL_STORAGE`.
```
Comment (ehanson8) [inline README.md:48]: This is very convenient!

File: abdiff/core/download_input_files.py (hunk 6 above)
Comment (ghukill) [reply abdiff/core/download_input_files.py:43]: > Hmm, I agree that in the case of large runs, it would be hard to parse through the varying logs to identify files that were already in the S3 bucket (and downloads were skipped for). 🤔 I'm good with the change proposed above if @ehanson8 agrees with the change as well!

I'm more concerned with counting to be honest.  I'm fully preparing to pass 1.6k input files to this, and I imagine the downloading may take.... 2-3 hours?  I think we need something like:

```
...
Input file 422 / 1630: 'alma-2024-10-01-full....xml' available locally for transformation.
Input file 423 / 1630: 'alma-2024-10-02-full....xml' available locally for transformation.
...
```

in the output to know how far along it is.  Of course, the _second_ time I run it, that list would sail by as those files are already present in Minio.  ABDiff is chatty, and I think that's okay; it'd just barf out 1.6k lines informing me that files are locally available for use.  But the first time, I think the incrementing progress is super helpful.

Which brings up your next comment:

> Outside the scope of this PR: It feels like a CLI command that informs / warns / reminds user to clean up their storage in the case of very large runs, but it seems like that CLI command would only know which directory to check via the MINIO_S3_LOCAL_STORAGE environment variable...

I would definitely lean into the idea this is "extras" functionality, and ABDiff has no responsibility to inform the user that they may have 50gb - 75gb in their minio server.  Open to pushback here, but data that big, it's hard _not_ to remember it's on disk!

File: abdiff/core/download_input_files.py
Comment (jonavellecuerdo) [reply abdiff/core/download_input_files.py:1]: That sounds fine with me! Good with you, @ehanson8 ?

```text
└── abdiff/
    └── extras/
        └── minio/
            ├── docker-compose.yaml
            └── download_input_files.py
```

File: abdiff/core/download_input_files.py (hunk 8 above)
Comment (ehanson8) [reply abdiff/core/download_input_files.py:15]: Feels like a ticket to me unless very simple to implement

File: abdiff/core/download_input_files.py (hunk 6 above)
Comment (ehanson8) [reply abdiff/core/download_input_files.py:43]: I agree with @ghukill 's proposed changes!

**Review by ehanson8 — COMMENTED** [review]
Looks great, amazing how simple this is to implement! I second Graham's requested changes

File: abdiff/core/download_input_files.py
Comment (ehanson8) [reply abdiff/core/download_input_files.py:1]: Yes!

File: Makefile (hunk 9 above)
Comment (jonavellecuerdo) [reply Makefile:64]: Yep! Needed it. 😅 Updated!

File: abdiff/helpers/minio/docker-compose.yaml
Comment (jonavellecuerdo) [reply abdiff/helpers/minio/docker-compose.yaml:1]: Updated.

File: abdiff/core/download_input_files.py (hunk 6 above)
Comment (jonavellecuerdo) [reply abdiff/core/download_input_files.py]: Please see the latest commit for updates to improve logging. The logs shown below depict a `rundiff` command that involves 2 valid input files and 1 invalid input file that results in a `RuntimeError`. Let me know what both of you think of the logged messages: 

```text
2024-11-07 11:54:38,856 INFO abdiff.cli.main() line 44: Logger 'root' configured with level=DEBUG
2024-11-07 11:54:38,856 INFO abdiff.cli.main() line 45: Running process
2024-11-07 11:54:38,859 INFO abdiff.core.init_run.init_run() line 24: Run directory created: output/jobs/test-minio/runs/2024-11-07_16-54-38
2024-11-07 11:54:39,030 INFO abdiff.extras.minio.download_input_files.download_input_files() line 49: Input file: 1 / 3: 's3://timdex-extract-dev-222053980223/libguides/libguides-2024-06-03-full-extracted-records-to-index.xml' available locally for transformation.
2024-11-07 11:54:39,034 INFO abdiff.extras.minio.download_input_files.download_input_files() line 49: Input file: 2 / 3: 's3://timdex-extract-dev-222053980223/dspace/dspace-2024-03-06-daily-extracted-records-to-index.xml' available locally for transformation.
2024-11-07 11:54:41,359 INFO abdiff.extras.minio.download_input_files.download_input_files() line 55: Input file: 3 / 3: 's3://timdex-extract-dev-222053980223/libguides/libguides-2024-11-07-daily-extracted-records-to-index.xml' failed to download.
2024-11-07 11:54:41,359 INFO abdiff.extras.minio.download_input_files.download_input_files() line 59: Available input files: 2, missing input files: 1.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/.local/share/virtualenvs/transmogrifier-ab-diff-3LIAwCT6/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jcuerdo/Documents/repos/transmogrifier-ab-diff/abdiff/cli.py", line 175, in run_diff
    download_input_files(input_files_list)
  File "/Users/jcuerdo/Documents/repos/transmogrifier-ab-diff/abdiff/extras/minio/download_input_files.py", line 64, in download_input_files
    raise RuntimeError(  # noqa: TRY003
RuntimeError: 1 input file(s) failed to download.
```

This change also led @ghukill and I to push back the work to [support concurrent file downloads](https://mitlibraries.atlassian.net/browse/TIMX-386) to give us more time to think about how we can maintain effective logging (i.e., concurrent downloads may result in disordered 'counts' in log messages).

File: abdiff/core/download_input_files.py (hunk 8 above)
Comment (jonavellecuerdo) [reply abdiff/core/download_input_files.py]: Created ticket [TIMX-386](https://mitlibraries.atlassian.net/browse/TIMX-386). For additional context for this decision, see https://github.com/MITLibraries/transmogrifier-ab-diff/pull/49#discussion_r1833038545.

File: abdiff/extras/minio/download_input_files.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+                "failed to download."
+            )
+    logger.info(
+        f"Available input files: {success_count}, missing input files: {fail_count}."
+    )
+
+    if fail_count > 0:
+        raise RuntimeError(  # noqa: TRY003
+            f"{fail_count} input file(s) failed to download."
+        )
```
Comment (ghukill) [inline abdiff/extras/minio/download_input_files.py:66]: This looks great!  I like the consistent logging structure, the final tally, and the exception raised if any failures.  This will be helpful for debugging large amounts of files to download, if anything goes wrong.

**Review by ehanson8 — APPROVED** [review]
Great work!
